pkg/bootconfig/zipconfig.go

```diff
@@ -0,0 +1,130 @@
+package bootconfig
+
+import (
+	"archive/zip"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+
+	"github.com/systemboot/systemboot/pkg/crypto"
+	"golang.org/x/crypto/ed25519"
+)
+
+// memoryZipReader is used to unpack a zip file from a byte sequence in memory.
+type memoryZipReader struct {
+	Content []byte
+}
+
+func (r *memoryZipReader) ReadAt(p []byte, offset int64) (n int, err error) {
+	cLen := int64(len(r.Content))
+	if offset > cLen {
+		return 0, io.EOF
+	}
+	if cLen-offset >= int64(len(p)) {
+		n = len(p)
+		err = nil
+	} else {
+		err = io.EOF
+		n = int(int64(cLen) - offset)
+	}
+	copy(p, r.Content[offset:int(offset)+n])
+	return n, err
+}
+
+// FromZip tries to extract a boot configuration from a ZIP file after verifying
+// its signature with the provided public key file. The signature is expected to
+// be appended to the ZIP file and have fixed length `ed25519.SignatureSize` .
+// The returned string argument is the temporary directory where the files were
+// extracted, if successful.
+// No decoder (e.g. JSON, ZIP) or other function parsing the input file is called
+// before verifying the signature.
+func FromZip(filename string, pubkeyfile *string) (*Manifest, string, error) {
```

Comment:
This seems to do many things, do we have a usecase for extracting the whole zip without parsing the manifest, or parsing the manifest without extracting the whole zip ? If so we may want a manifest-extract-only function and an extract-everything function ?

Reply:
fair point. I don't think we have a use case to do unpacking without manifest extraction and optional signature verification. However I could still move the ZIP unpacking in its own `Unpack` method, and wrap it within `FromZip` (that will call `Unpack` and look for a valid manifest in the output directory). How does that sound?

Reply:
Definitely makes sense to not reextract just to find the manifest if we've already extracted before. You'd need some heavier redesign to be able to just reuse the previously extracted image (at least store the location in tmpfs) and since we don't have a usecase right now let's add the complexity later

Reply:
thanks. If the other updates are good can you please re-approve?